### PR TITLE
Solana address table lookup

### PR DESCRIPTION
**Pull request by barnjamin**
Resubmit of #129 

a (devnet) checkable transaction for the account ordering is here: https://explorer.solana.com/tx/2nnN1oinF53chzse4UQA4fHAzFmkynrFZWF4evdjuRbYzE6To3Q4qUvWsKzJ4GSJBzFbM8xjYgXfteapPGc2fbGx?cluster=devnet

for this transaction we're producing the correct list of accounts in order:
```
[2023-10-26 15:16:37.073 -0400] [info] [solana] processing 1 transactions
4q2wPZMys1zCoAVpNmhgmofb6YM9MqLXmV25LdtEMAf9
vQkE51MXJiwqtbwf562XWChNKZTgh6L2jHPpupoCKjS
5Ng1zHMKVDSVd17TTJJBPzgk4aFptE4rg5tmW2Ekk4Zp
A5DjbpHjGTcWkCYfz976adGCBMJJQN1vLib4izrUNBnY
G94RRLkVDzfrqZ8gQSYGZY6FW6vADrieQ9U3oTXGen3W
GP5mKU4d3zKeGS61281A6h5yB1D3AcKjaNoGSCgtrDhb
ComputeBudget111111111111111111111111111111
TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA
2UGDsGfymRHpzhRq5dxgCaJdvdG9J5xnLxTAqAS6usLC
3u8hJUVTA4jH1wYAyUur7FFZVQ8H635K3tSHHF4ssjQ5
5DncnqicaHDZTMfkcfzKaYP5XzD5D9jg3PGNTT5J1Qg7
7oZWSW5vZvT8AeHn3CtS7RPVBDsw5UR8hLburyHz92yd
ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL
CZXxEDtjzuWMQonx35T4hYjkxA9WKwgkn4Uq172t4fAm
CtEV3dA9t79gc9PJKSkH2WehSbR28jVdMxsaQmvwDmMg
DZnkkTmCiFWfYTfT41X3Rd1kDgozqzxWaHqsw6W4x2oe
9QzqZZvhxoHzXbNY9y2hyAUfJUzDwyDb7fbDs9RXwH3
4yttKWzRoNYS2HekxDfcZYmfQqnVWpKiJ8eydYRuFRgs
4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU
68DjnBuZ6UtM6dGoTGhu2rqV5ZSowsPGgv2AWD1xuGB4
6bi4JGDoRwUs9TYBuvoA7dUVyikTJDrJsJU1ew6KVLiu
7s3a1ycs16d6SNDumaRtjcoyMaTDZPavzgsmS3uUZYWX
11111111111111111111111111111111
SysvarC1ock11111111111111111111111111111111
SysvarRent111111111111111111111111111111111
3VFdJkFuzrcwCwdxhKRETGxrDtUVAipNmYcLvRBDcQeH
8PFZNjn19BBYVHNp4H31bEW7eAmu78Yf2RKV8EeA461K
H9pUTqZoRyFdaedRezhykA1aTMq7vbqRHYVhpHZK2QbC
{
  '253807610/2023-10-26T11:59:03.000Z': [],
  '253807609/2023-10-26T11:59:02.000Z': [
    'QbKEDNkXuC9W6EExUkJwrkM544URLfznSFHvWVbfjNdQxkT92C8ry4meAUo9n7vF7Px7n8EuSzfGZwE8RSkbHan:1/3b26409f8aaded3f5ddca184695aa6a0fa829b0c85caf84856324896d214ca98/25595'
  ]
}
```

### CODE REVIEW

File: watcher/src/watchers/__tests__/SolanaWatcher.test.ts [hunk 1]
```diff
@@ -32,21 +32,21 @@ test('getMessagesForBlocks - fromSlot is skipped slot', async () => {
...
   expect(Object.keys(messages).length).toBe(1);
-  expect(messages).toMatchObject({ '171774032/2023-01-10T13:36:38.000Z': [] });
+  expect(messages).toMatchObject({ '171774032/2023-01-10T13:36:39.000Z': [] });
 });
 
 test('getMessagesForBlocks - toSlot is skipped slot', async () => {
   const watcher = new SolanaWatcher();
   const messages = await watcher.getMessagesForBlocks(171774023, 171774025);
-  expect(messages).toMatchObject({ '171774023/2023-01-10T13:36:34.000Z': [] });
+  expect(messages).toMatchObject({ '171774025/2023-01-10T13:36:34.000Z': [] });
```
Comment (barnjamin) [inline watcher/src/watchers/__tests__/SolanaWatcher.test.ts:41]: Note this represents a behavior change, we now return the last entry.

This was not intentional but I think it makes sense

File: watcher/src/watchers/SolanaWatcher.ts [hunk 2]
```diff
@@ -129,37 +154,66 @@ export class SolanaWatcher extends Watcher {
...
+
         const whInstructions = innerInstructions
           .concat(instructions)
           .filter((i) => i.programIdIndex === programIdIndex);
+
         for (const instruction of whInstructions) {
           // skip if not postMessage instruction
           const instructionId = instruction.data;
-          if (instructionId[0] !== 0x01) continue;
+          if (instructionId[0] !== 0x08 && instructionId[0] !== 0x01) continue;
```
Comment (barnjamin) [inline watcher/src/watchers/SolanaWatcher.ts:198]: Note: to pick up the example transaction from the reason for this pr in the first place I had to add 0x08 as a possible instruction id

File: watcher/src/watchers/SolanaWatcher.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import {
   PublicKey,
   SolanaJSONRPCError,
   VersionedBlockResponse,
+  VersionedTransactionResponse,
```
Comment (panoel) [inline watcher/src/watchers/SolanaWatcher.ts:10]: I don't see this used anywhere?

File: watcher/src/watchers/SolanaWatcher.ts [hunk 4]
```diff
@@ -31,71 +33,92 @@ export class SolanaWatcher extends Watcher {
   // of the WH transactions within that range is handled internally below.
   maximumBatchSize = 100_000;
 
-  constructor() {
+  connection: Connection | undefined;
+
+  constructor(rpc: string = RPCS_BY_CHAIN.solana!, contract: string = WORMHOLE_PROGRAM_ID) {
```
Comment (panoel) [inline watcher/src/watchers/SolanaWatcher.ts:38]: Why did you choose to add parameters to the constructor when they are never changed?

File: watcher/src/watchers/SolanaWatcher.ts [hunk 5]
```diff
@@ -31,71 +33,92 @@ export class SolanaWatcher extends Watcher {
...
     super('solana');
-    this.rpc = RPCS_BY_CHAIN.solana!;
+    this.rpc = rpc;
+    this.programId = contract;
   }
 
-  async getFinalizedBlockNumber(): Promise<number> {
-    const connection = new Connection(this.rpc, COMMITMENT);
-    return connection.getSlot();
+  getConnnection(): Connection {
```
Comment (panoel) [inline watcher/src/watchers/SolanaWatcher.ts:44]: Maybe remove one 'n'? (getConnnnnnnnnnnection())

File: watcher/src/watchers/SolanaWatcher.ts [hunk 6]
```diff
@@ -129,37 +154,66 @@ export class SolanaWatcher extends Watcher {
...
+            const lookupTableAccount = await this.getConnnection()
+              .getAddressLookupTable(atl.accountKey)
+              .then((res) => res.value);
+
+            if (!lookupTableAccount)
+              throw new Error('lookupTableAccount is null, cant resolve addresses');
+
+            // Important to return the addresses in the order they're specified in the
+            // address table lookup object. Note writable comes first, then readable.
+            // TODO: check if this is broken by multiple addressTableLookups in a single message
```
Comment (panoel) [inline watcher/src/watchers/SolanaWatcher.ts:174]: Did you answer this question?

**Review by panoel — APPROVED** [review]

File: watcher/src/watchers/SolanaWatcher.ts (hunk 4 above)
Comment (barnjamin) [reply watcher/src/watchers/SolanaWatcher.ts:38]: I was changing them in the script but not anymore, will remove

File: watcher/src/watchers/SolanaWatcher.ts (hunk 6 above)
Comment (barnjamin) [reply watcher/src/watchers/SolanaWatcher.ts:172]: No,  I forgot about this, need to dig in and remember what the heck this means

Comment (barnjamin) [reply watcher/src/watchers/SolanaWatcher.ts]: Yea this woulda been buggy, I think I've got the correct logic now but need to find a tx to test against

**Review by panoel — APPROVED** [review]

**Review by panoel — APPROVED** [review]

**Review by panoel — APPROVED** [review]

**Pull request merged**